apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,272 @@
+        return this.UNDEFINED;
+      } else {
+        customMarshalValue = obj.data[name];
+      }
+    } else {
+      var hasProperty = false;
+      if (!obj.isPrimitive) {
+        hasProperty = super.hasProperty(obj, name);
+      }
+      if (!hasProperty &&
+          obj === this.globalScope &&
+          !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+          !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)) {
+        customMarshalValue = nativeParent[name];
+      } else {
+        return super.getProperty(obj, name);
+      }
+    }
+    var type = typeof customMarshalValue;
+    if (type === 'number' || type === 'boolean' || type === 'string' ||
+        type === 'undefined' || customMarshalValue === null) {
+      return this.createPrimitive(customMarshalValue);
+    } else {
+      return codegen.marshalNativeToInterpreter(this, customMarshalValue, obj.data);
+    }
+  }
+
+  /**
+   * Wrapper to Interpreter's hasProperty (extended for custom marshaling)
+   *
+   * Does the named property exist on a data object.
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @return {boolean} True if property exists.
+   * @override
+   */
+  hasProperty(obj, name) {
+    name = name.toString();
+    var nativeParent;
+    if (obj.isCustomMarshal) {
+      if (this.shouldBlockCustomMarshalling_(name, obj)) {
+        return false;
+      } else {
+        return name in obj.data;
+      }
+    } else {
+      var hasProperty = super.hasProperty(obj, name);
+      if (!hasProperty &&
+          obj === this.globalScope &&
+          !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+          !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)) {
+        return true;
+      } else {
+        return hasProperty;
+      }
+    }
+  }
+
+  /**
+   * Wrapper to Interpreter's setProperty (extended for custom marshaling)
+   *
+   * Set a property value on a data object.
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @param {*} value New property value.
+   * @param {boolean} opt_fixed Unchangeable property if true.
+   * @param {boolean} opt_nonenum Non-enumerable property if true.
+   * @override
+   */
+  setProperty(
+    obj,
+    name,
+    value,
+    opt_fixed,
+    opt_nonenum
+  ) {
+    name = name.toString();
+    var nativeParent;
+    if (obj.isCustomMarshal) {
+      if (!this.shouldBlockCustomMarshalling_(name, obj)) {
+        obj.data[name] = codegen.marshalInterpreterToNative(this, value);
+      }
+    } else {
+      var hasProperty = false;
+      if (!obj.isPrimitive) {
+        hasProperty = super.hasProperty(obj, name);
+      }
+      if (!hasProperty &&
+          obj === this.globalScope &&
+          !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+          !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)) {
```

Comment:
This complex conditional gets reused in `getProperty`, `hasProperty` and `setProperty`.  Let's extract a helper.